task-sdk/src/airflow/sdk/execution_time/context.py

```diff
@@ -142,8 +143,10 @@ def _get_connection(conn_id: str) -> Connection:
     from airflow.sdk.execution_time.comms import ErrorResponse, GetConnection
     from airflow.sdk.execution_time.task_runner import SUPERVISOR_COMMS
 
-    SUPERVISOR_COMMS.send_request(log=log, msg=GetConnection(conn_id=conn_id))
-    msg = SUPERVISOR_COMMS.get_message()
+    with SUPERVISOR_COMMS.lock:
```

Comment:
Why do we need `.lock`?

Reply:
That was a discussion with Ash where we felt it was needed for the different Triggers to not use the communication channel at the same time and receive / send wrong messages. (because they are all async).

Therefore when a trigger is trying to send/receive something from the supervisor, it will block the comms so it does not interfere with other triggers.

Reply:
It's probably worth adding a comment here saying why we need this
```suggestion
    # Since Triggers can hit this code path via `sync_to_async` we need to make sure that we
    # "atomically" send a request and get the response to that back so that two triggers don't
    # end up interleaving requests and end up with a possible race condition where the wrong
    # trigger reads the response.
    with SUPERVISOR_COMMS.lock:
```

Reply:
Good idea, done to the 4 places. We might want to refactor that into a method that can `read_write` => atomic_read_write and therefore the comment can just be there and the lock to one place.

Reply:
Yup, there's also #46426